rfd/0022-ssh-agent-forwarding.md

```diff
@@ -0,0 +1,205 @@
+   └─────────┘          └─────────────┘          └──────────────┘          └─────────────┘
+```
+
+The above diagram leaves out a _lot_ of detail, but the overall process should be clear
+enough for this discussion.
+
+### What's the issue?
+
+An OpenSSH user can forward their Key Agent to the remote machine using the `-A` flag on
+the `ssh` command line. This allows the user to use arbitrary keys stored on their _local_
+machine to authenticate against a "Third Machine" from inside their session on the Remote
+Machine.
+
+When running `tsh ssh`, a user may have _two_ independent key agents running: their own
+(which we will call the _System Key Agent_), and an ephemeral one inside the `tsh` process
+itself (the _Teleport Key Agent_). The Teleport Key Agent is populated with the contents
+of the user's `.tsh` directory.
+
+The `tsh ssh` client _also_ offers a `-A` option to forward a Key Agent to the Remote
+Machine, but it will only ever forward the Teleport Key Agent, not the System Key Agent.
+This is surprising behaviour to a user accustomed to OpenSSH. The user finds that keys
+they expected to be available to them in the session on the Remote Machine are not
+available.
+
+The tricky part is that _both_ of these behaviours are sensible, depending on the user's
+expectations. Any solution will have to accommodate both behaviours.
+
+## Details
+
+### Proposed Solution
+
+**NB:** This proposal is strictly about how a Key Agent is forwarded to a remote 
+machine by `tsh ssh`. It does _not_ change how `tsh` authenticates against a `teleport`
+auth service.
+
+Behind the scenes, both the OpenSSH client & `tsh ssh` treat the `-A` flag as shorthand
+for setting the SSH config value `ForwardAgent` to `yes`.
+
+Looking at the `man` page for `ssh_config(5)`, we can see the definitive list of values
+that `ForwardAgent` can take, and their meaning:
+
+> **ForwardAgent**
+>
+>   Specifies whether the connection to the authentication agent (if any) will be forwarded
+>   to the remote machine.  The argument may be `yes`, `no` (the default), an explicit path 
+>   to an agent socket or the name of an environment variable (beginning with ‘$’) in which 
+>   to find the path.
+
+The `tsh ssh` client already supports the `yes` and `no` options, where `yes` means
+forwarding the Teleport Key Agent. To allow the user to forward their own Key Agent, _and_
+to bring `tsh ssh`'s behaviour more in line with the OpenSSH client, I propose the following
+values be allowed for the `ForwardAgent`option:
+
+
+| Value          | Interpretation                                           |
+|----------------|----------------------------------------------------------|
+| `no` (default) | `tsh ssh` will not forward _any_ Key Agent               |
+| `yes`          | `tsh ssh` will forward the System Key Agent (if present) |
+| `local`        | `tsh ssh` will forward the Teleport Key Agent            |
```

Comment:
I know I'm a bit too late, but I was wondering if wouldn't make sense to keep the openssh (yes/no) and use an external configuration (like an environment variable or a config file) to control weather or not to export the local agent. 

For the user, `-o ForwardAgent local` is way more verbose, introduce the precedence issue you mentioned and seem a bit counter-intuitive.